File: TouchDriver/simInput.m
//
//  simInput.m
//  TouchDriver
//
//  Created by Daniel Prilik on 2017-08-17.
//

//---------------------------------------------------------------------------
// Simulate Mouse Events
//---------------------------------------------------------------------------

#include "utils.h"

void simulateClick(int x, int y, ButtonState button) {
    if (button == BTN_DOWN) {
        CGEventRef mouse_press = CGEventCreateMouseEvent(NULL,
                                                         kCGEventLeftMouseDown,
                                                         CGPointMake(x, y),
                                                         kCGMouseButtonLeft);
        //CGEventSetIntegerValueField(mouse_press, kCGMouseEventNumber, eventNumber);
        CGEventPost(kCGHIDEventTap, mouse_press);
        CFRelease(mouse_press);
        //eventNumber++;
    }
    else if (button == BTN_UP) {
        CGEventRef mouse_release = CGEventCreateMouseEvent(NULL,
                                                           kCGEventLeftMouseUp,
                                                           CGPointMake(x, y),
                                                           kCGMouseButtonLeft);
        //CGEventSetIntegerValueField(mouse_release, kCGMouseEventNumber, eventNumber);
        CGEventPost(kCGHIDEventTap, mouse_release);
        CFRelease(mouse_release);
        //eventNumber++;
    }
    else if (button == BTN_RIGHT) {
        CGEventRef mouse_right = CGEventCreateMouseEvent(NULL,
                                                         kCGEventRightMouseDown,
                                                         CGPointMake(x, y),
                                                         kCGMouseButtonRight);
        //CGEventSetIntegerValueField(mouse_press, kCGMouseEventNumber, eventNumber);
        CGEventPost(kCGHIDEventTap, mouse_right);
        CGEventSetType(mouse_right, kCGEventRightMouseUp);
        CGEventPost(kCGHIDEventTap, mouse_right);
        CFRelease(mouse_right);
        //eventNumber++;
    }
    else if (button == BTN_2_CLICK)
    {
        CGEventRef mouse_double = CGEventCreateMouseEvent(NULL,
                                                          kCGEventLeftMouseDown,
                                                          CGPointMake(x, y),
                                                          kCGMouseButtonLeft);
        CGEventSetIntegerValueField(mouse_double, kCGMouseEventClickState, 2);

        CGEventPost(kCGHIDEventTap, mouse_double);
        CGEventSetType(mouse_double, kCGEventLeftMouseUp);
        CGEventPost(kCGHIDEventTap, mouse_double);
        CGEventSetType(mouse_double, kCGEventLeftMouseDown);
        CGEventPost(kCGHIDEventTap, mouse_double);
        CGEventSetType(mouse_double, kCGEventLeftMouseUp);
        CGEventPost(kCGHIDEventTap, mouse_double);
        CFRelease(mouse_double);
    }

    if (button == BTN_NO_CHANGE) {
        CGEventRef move = CGEventCreateMouseEvent(NULL,
                                                  kCGEventLeftMouseDragged,
                                                  CGPointMake(x, y),
                                                  kCGMouseButtonLeft);
        //CGEventSetIntegerValueField(move, kCGMouseEventNumber, eventNumber);
        CGEventPost(kCGHIDEventTap, move);
        CFRelease(move);
        //eventNumber++;
    }

    if (button == BTN_MOVE) {
        CGEventRef move = CGEventCreateMouseEvent(NULL,
                                                  kCGEventMouseMoved,
                                                  CGPointMake(x, y),
                                                  kCGMouseButtonLeft);
        //CGEventSetIntegerValueField(move, kCGMouseEventNumber, eventNumber);
        CGEventPost(kCGHIDEventTap, move);
        CFRelease(move);
        //eventNumber++;
    }
}

#import <ScriptingBridge/ScriptingBridge.h>
#import "SystemEvents.h"

// https://stackoverflow.com/questions/29032726/open-notification-center-programmatically-on-os-x
void openNotificationCenter() {
    static bool firstrun = true;
    static SystemEventsMenuBarItem *item = nil;

    if (firstrun) {
        SystemEventsApplication *systemEventsApp = (SystemEventsApplication *)[[SBApplication alloc] initWithBundleIdentifier:@"com.apple.systemevents"];
        SystemEventsApplicationProcess *sysUIServer = [systemEventsApp.applicationProcesses objectWithName:@"SystemUIServer"];

        for (SystemEventsMenuBar *menuBar in sysUIServer.menuBars) {
            item = [menuBar.menuBarItems objectWithName:@"Notification Center"];
            if (item != nil && [item.name isEqualToString:@"Notification Center"])
                break;
        }
    }

    firstrun = false;
    [item clickAt:nil];
}

File: TouchDriver/simInput.swift
//
//  simInput.swift
//  TouchDriver
//
//  Created by Daniel Prilik on 2017-08-21.
//

import Foundation

func simulateClick(x: Int, y: Int, button: ButtonState) {
    let point = CGPoint(x: x, y: y)

    switch button {
    case BTN_DOWN:
        let mouse_press = CGEvent(mouseEventSource: nil,
                                  mouseType: CGEventType.leftMouseDown,
                                  mouseCursorPosition: point,
                                  mouseButton: CGMouseButton.left)

        mouse_press?.post(tap: CGEventTapLocation.cghidEventTap)
    case BTN_UP:
        let mouse_release = CGEvent(mouseEventSource: nil,
                                    mouseType: CGEventType.leftMouseUp,
                                    mouseCursorPosition: point,
                                    mouseButton: CGMouseButton.left)

        mouse_release?.post(tap: CGEventTapLocation.cghidEventTap)
    case BTN_RIGHT:
        let mouse_right = CGEvent(mouseEventSource: nil,
                                  mouseType: CGEventType.rightMouseDown,
                                  mouseCursorPosition: point,
                                  mouseButton: CGMouseButton.right)

        mouse_right?.post(tap: CGEventTapLocation.cghidEventTap)
        mouse_right?.type = CGEventType.rightMouseUp
        mouse_right?.post(tap: CGEventTapLocation.cghidEventTap)

    case BTN_2_CLICK:
        let mouse_double = CGEvent(mouseEventSource: nil,
                                   mouseType: CGEventType.leftMouseDown,
                                   mouseCursorPosition: point,
                                   mouseButton: CGMouseButton.left)
        mouse_double?.setIntegerValueField(CGEventField.mouseEventClickState, value: 2)

        mouse_double?.post(tap: CGEventTapLocation.cghidEventTap)
        mouse_double?.type = CGEventType.leftMouseUp
        mouse_double?.post(tap: CGEventTapLocation.cghidEventTap)
    case BTN_NO_CHANGE:
        let move = CGEvent(mouseEventSource: nil,
                           mouseType: CGEventType.leftMouseDragged,
                           mouseCursorPosition: point,
                           mouseButton: CGMouseButton.left)
        move?.post(tap: CGEventTapLocation.cghidEventTap)
    case BTN_MOVE:
        break
    case BTN_DUMMY:
        break
    default:
        break
    }
}

let applescripts: [String: NSAppleScript?] = [
    "notification_center": NSAppleScript(source: """
        tell application "System Events"
            click menu bar item "Notification Center" of menu bar 1 of application process "SystemUIServer"
        end tell
    """)
]

func openNotificationCenter() {
    var error: NSDictionary?
    applescripts["notification_center"]!?.executeAndReturnError(&error)
}

